### PR TITLE
[SIM_FLUTTER-106] [FE] Update UI in withdrawal

### PR DESCRIPTION
**Pull request by CaneteJoash**
#### LINKS
- backlog link: https://framgiaph.backlog.com/board/SIM_FLUTTER?selectedIssueKey=SIM_FLUTTER-106
- Figma Link: https://www.figma.com/file/uzAwGmOCLvGbReXshnT3UN/SAVINGS-APP?type=design&node-id=871-1277&mode=design&t=WuF833sezrgn3mgX-0
- slack Link: https://sun-philippine.slack.com/archives/C059L7XAXFB/p1688717415398919

#### DOD
- [X] Update UI in withdrawal

#### Note
- Category is still static

#### Screenshots 
image
image

### CODE REVIEW

**Review by Dbulawan13 — DISMISSED** [review]
Approve PR #55 

Working:
- Able to Update UI for withdrawal
- Dropdown button display correct categories based on Figma

https://github.com/framgia/sph-flutter/assets/86758474/3cb298f4-744e-450f-83fe-eb27ee37e4b1

Testing hours: 0.08
Comment: 0
Iteration: 0

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 1]
```diff
@@ -4,23 +4,37 @@ import 'package:frontend/src/models/transaction.dart';
 import 'package:frontend/src/services/transaction_service.dart';
 
 List<String> transactionTypes = ['Credit', 'Dept', 'Transfer'];
+List<String> transactionCategory = [
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: move category and types to /enums/transaction_enum.dart

transfer enums under /models/transaction.dart as well

Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: change implementation to enum

`Foods/Drinks` is not valid as enum because of the `slash`
think of another term to encapsulate both and conditionally render `Foods/Drinks` based on that new enum value

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 2]
```diff
@@ -4,23 +4,37 @@ import 'package:frontend/src/models/transaction.dart';
...
+  'Transportation',
+  'Housing/Billings',
+  'Savings',
+  'Miscellaneous'
+];
 
 class TransactionController extends GetxController {
   final RxString _selectedTransactionType = transactionTypes[0].obs;
 
+  final RxString _selectedTransactionCategory = transactionCategory[0].obs;
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: compress variables here, no space

File: frontend/lib/src/controllers/transaction_controller.dart (hunk 1 above)
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: all transaction categories must be included in the new enum

for getting the values only meant for, for example Transfer type
i tried this 
image

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #55

review time: 15mins
comments: 4
iteration: 1st review

File: frontend/lib/src/enums/transaction_enum.dart [hunk 3]
```diff
@@ -0,0 +1,59 @@
+// ignore_for_file: constant_identifier_names
+
+enum TransactionTypes { CREDIT, DEPT, TRANSFER }
+
+enum Category {
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: category for what

File: frontend/lib/src/enums/transaction_enum.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  TRANSPORTATION,
+  BILLS,
+  SAVINGS,
+  SALARY,
+  SENDER,
+  RECIPIENT,
+  MISC,
+}
+
+List<String> creditCategory() {
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: check if this can be variable instaed of function

Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]:
```suggestion
List<String> creditCategories() {
```

File: frontend/lib/src/enums/transaction_enum.dart [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return 'Savings';
+      case Category.MISC:
+        return 'Miscellaneous';
+      default:
+        return '';
+    }
+  }).toList();
+}
+
+String stringToEnum(input) {
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: mention transaction here, this is very ambiguous

File: frontend/lib/src/enums/transaction_enum.dart [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      return Category.TRANSPORTATION.name;
+    case 'Housing/Billings':
+      return Category.BILLS.name;
+    case 'Savings':
+      return 'Savings';
+    case 'Miscellaneous':
+      return Category.MISC.name;
+    default:
+      return Category.MISC.name;
+  }
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: this doesnt return enum

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 7]
```diff
@@ -18,6 +19,13 @@ import 'package:frontend/src/features/dashboard/components/account_card.dart';
 class TransactionHistory extends StatelessWidget {
   const TransactionHistory({super.key});
 
+  String capitalizeFirstLetter(String input) {
+    if (input.isEmpty) {
+      return input;
+    }
+    return input[0].toUpperCase() + input.substring(1).toLowerCase();
+  }
+
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: move to /helper

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 8]
```diff
@@ -49,12 +57,30 @@ class TransactionHistory extends StatelessWidget {
...
-                      selectedValue: controller.selectedTransactionType,
+                      items: TransactionTypes.values
+                          .map(
+                            (type) => capitalizeFirstLetter(
+                              type.toString().split('.').last,
+                            ),
+                          )
+                          .toList(),
+                      selectedValue: capitalizeFirstLetter(
+                        controller.selectedTransactionType
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: use enum.name here

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 9]
```diff
@@ -49,12 +57,30 @@ class TransactionHistory extends StatelessWidget {
                   children: [
                     Dropdown(
                       labelText: 'Filter transaction by type',
-                      items: transactionTypes,
-                      selectedValue: controller.selectedTransactionType,
+                      items: TransactionTypes.values
+                          .map(
+                            (type) => capitalizeFirstLetter(
+                              type.toString().split('.').last,
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: use .name

File: frontend/lib/src/features/transaction_history/transaction_history.dart [hunk 10]
```diff
@@ -49,12 +57,30 @@ class TransactionHistory extends StatelessWidget {
...
+                            .last,
+                      ),
                       onChanged: (value) {
                         controller.setSelectedTransactionType =
-                            value.toString();
-                        debugPrint(controller.selectedTransactionType);
+                            TransactionTypes.values.firstWhere(
+                          (type) =>
+                              capitalizeFirstLetter(
+                                type.toString().split('.').last,
```
Comment (jstephend-sun) [inline frontend/lib/src/features/transaction_history/transaction_history.dart]: use .name

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 11]
```diff
@@ -160,6 +162,26 @@ class TransactionComponent extends StatelessWidget {
...
                 ],
+                if (type == TransactionTypes.CREDIT) ...[
+                  Dropdown(
+                    labelText: 'Category',
+                    items: creditCategory(),
+                    selectedValue:
+                        transactionController.selectedTransactionCategory,
+                    onChanged: (value) {
+                      transactionController.setSelectedTransactionCategory =
+                          value.toString();
```
Comment (jstephend-sun) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: isnt the value already a string

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 12]
```diff
@@ -1,25 +1,31 @@
...
 
   DateTime get selectedTransactionDate => _selectedTransactionDate.value;
 
   bool get transactionSubmitEnabled => _transactionSubmitEnabled.value;
 
-  set setSelectedTransactionType(String newValue) =>
+  set setSelectedTransactionType(TransactionTypes newValue) =>
       _selectedTransactionType.value = newValue;
 
+  set setSelectedTransactionCategory(String newValue) =>
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]: this should be enum

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes ##55

total review time: 30mins 
comments: 4 + 11
iteration: 2nd review

**Review by Dbulawan13 — DISMISSED** [review]

File: frontend/lib/src/enums/transaction_enum.dart (hunk 3 above)
Comment (CaneteJoash) [reply frontend/lib/src/enums/transaction_enum.dart]: Sir Good afternoon, this category was from `/models/transaction.dart/` I just transfer it here in `/src/enums/transaction_enum.dart` and I believe this enum was used somewhere. so I didn't  change the name.

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 11 above)
Comment (CaneteJoash) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: I added `.toString` because A value of type 'Object?' can't be assigned to a variable of type 'String'.
Try changing the type of the variable, or casting the right-hand type to 'String'.dart[invalid_assignment](https://dart.dev/diagnostics/invalid_assignment)
Object? value
Type: Object?

File: frontend/lib/src/enums/transaction_enum.dart (hunk 3 above)
Comment (jstephend-sun) [reply frontend/lib/src/enums/transaction_enum.dart]: `Category` is ambiguous, change enum name

File: frontend/lib/src/controllers/transaction_controller.dart [hunk 13]
```diff
@@ -1,25 +1,33 @@
...
 import 'package:frontend/src/models/transaction.dart';
+import 'package:frontend/src/enums/transaction_enum.dart';
 import 'package:frontend/src/services/transaction_service.dart';
 
-List<String> transactionTypes = ['Credit', 'Dept', 'Transfer'];
-
 class TransactionController extends GetxController {
-  final RxString _selectedTransactionType = transactionTypes[0].obs;
+  final Rx<TransactionTypes> _selectedTransactionType =
+      TransactionTypes.values[0].obs;
```
Comment (jstephend-sun) [inline frontend/lib/src/controllers/transaction_controller.dart]:
```suggestion
     ` TransactionTypes.values[0].obs;`
```
simplify

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 14]
```diff
@@ -113,7 +115,7 @@ class TransactionComponent extends StatelessWidget {
     return SingleChildScrollView(
       padding: const EdgeInsets.fromLTRB(25, 30, 25, 0),
       child: SizedBox(
-        height: type == TransactionTypes.TRANSFER ? 450 : 350,
+        height: 450,
```
Comment (jstephend-sun) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: why was this removed? 

unnecessary code change

File: frontend/lib/src/enums/transaction_enum.dart [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    case TransactionCategories.SAVINGS:
+      return 'Savings';
+    case TransactionCategories.MISC:
+      return 'Miscellaneous';
+    default:
+      throw ArgumentError('Invalid credit transaction: $category');
+  }
+}
+
+TransactionCategories stringToEnum(String creditTransaction) {
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: still ambiguous, can use (enum)fromstring or someting like that, be descriptive

File: frontend/lib/src/enums/transaction_enum.dart [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    case 'Transportation':
+      return TransactionCategories.TRANSPORTATION;
+    case 'Housing/Billings':
+      return TransactionCategories.BILLS;
+    case 'Savings':
+      return TransactionCategories.SAVINGS;
+    case 'Miscellaneous':
+      return TransactionCategories.MISC;
+    default:
+      throw ArgumentError('Invalid credit transaction: $creditTransaction');
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: hmmm, default to misc for now

File: frontend/lib/src/enums/transaction_enum.dart [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    case TransactionCategories.BILLS:
+      return 'Housing/Billings';
+    case TransactionCategories.SAVINGS:
+      return 'Savings';
+    case TransactionCategories.MISC:
+      return 'Miscellaneous';
+    default:
+      throw ArgumentError('Invalid credit transaction: $category');
+  }
+}
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: remove, refactor as dart enum extension

default as "" empty string

image

File: frontend/lib/src/enums/transaction_enum.dart [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    case TransactionCategories.SAVINGS:
+      return 'Savings';
+    case TransactionCategories.MISC:
+      return 'Miscellaneous';
+    default:
+      return '';
+  }
+}).toList();
+
+String enumToString(TransactionCategories category) {
```
Comment (jstephend-sun) [inline frontend/lib/src/enums/transaction_enum.dart]: your function names are still very vague

**Review by jstephend-sun — CHANGES_REQUESTED** [review]
Requested Changes #55

total review time: 40mins
comments: 4 + 11 + 6
iteration: 3rd review

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 14 above)
Comment (jstephend-sun) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: needs rebase, rob's pr needs the height for this one

**Review by jstephend-sun — APPROVED** [review]
Approved #55

pls update actual hours 
1st review 07/12 10am
2nd review 07/12 5pm  (6hr gap)
3rd review 07/13 1pm (6hr + 2hr)

time from first review until approval 8hr
cc @Dbulawan13 

total review time: 40mins
comments: 21 (4 + 11 + 6)
iteration: 3rd review

**Review by Dbulawan13 — APPROVED** [review]
